packages/xo-server/src/xo-mixins/metadata-backups.js

```diff
@@ -0,0 +1,247 @@
+    if (schedule === undefined) {
+      throw new Error('backup job cannot run without a schedule')
+    }
+
+    const job: MetadataBackupJob = (job_: any)
+    const remoteIds = unboxIdsFromPattern(job.remotes)
+    if (remoteIds.length === 0) {
+      throw new Error('metadata backup job cannot run without remotes')
+    }
+
+    const poolIds = unboxIdsFromPattern(job.pools)
+    const isEmptyPools = poolIds.length === 0
+    if (!job.xoMetadata && isEmptyPools) {
+      throw new Error('no metadata mode found')
+    }
+
+    const app = this._app
+    const { retentionXoMetadata, retentionPoolMetadata } =
+      job?.settings[schedule.id] || {}
+
+    const scheduleDir = `${METADATA_BACKUP_DIR}/${schedule.id}`
+    const timestamp = safeDateFormat(Date.now())
+    const files = []
+    if (job.xoMetadata && retentionXoMetadata > 0) {
+      const dir = `${scheduleDir}/xo`
+      const fileName = `${dir}/${timestamp}.json`
+      const metaDataFileName = `${dir}/${timestamp}.metadata.json`
+
+      const data = await app.exportConfig()
+
+      files.push({
+        executeBackup: async handler => {
+          await Promise.all([
+            handler.outputFile(fileName, JSON.stringify(data, null, 2)),
+            handler.outputFile(
+              metaDataFileName,
+              JSON.stringify(
+                {
+                  jobId: job.id,
+                  scheduleId: schedule.id,
+                  timestamp,
+                },
+                null,
+                2
+              )
+            ),
+          ])
+          $defer.onFailure(async () => {
+            await Promise.all([
+              handler.unlink(fileName),
+              handler.unlink(metaDataFileName),
+            ])
+          })
+        },
+        dir,
+        retention: retentionXoMetadata - 1,
+      })
+    }
+    if (!isEmptyPools && retentionPoolMetadata > 0) {
+      const retention = retentionPoolMetadata - 1
+      files.push(
+        ...(await Promise.all(
+          poolIds.map(async id => {
+            const dir = `${scheduleDir}/${id}`
+            const fileName = `${dir}/${timestamp}.xml`
+            const metaDataFileName = `${dir}/${timestamp}.metadata.json`
+
+            const stream = await app.getXapi(id).getPoolMetadata(cancelToken)
+
+            return {
+              executeBackup: async handler => {
+                const [outputStream] = await Promise.all([
+                  handler.createOutputStream(fileName),
+                  handler.outputFile(
+                    metaDataFileName,
+                    JSON.stringify(
+                      {
+                        jobId: job.id,
+                        scheduleId: schedule.id,
+                        poolId: id,
+                        timestamp,
+                      },
+                      null,
+                      2
+                    )
+                  ),
+                ])
```

Comment:
What if `createOutputStream` or `outputFile` fails?

Reply:
No, it should be both or neither.

This is something that's not correctly handled in Backup NG and it's a big problem.

Reply:
That's better but the stream should also be destroyed if `outputFile` fails.